typing/typecore.ml

```diff
@@ -4426,8 +4426,7 @@ and type_let
               type_expect exp_env sexp (mk_expected pat.pat_type)))
       spat_sexp_list pat_slot_list in
   current_slot := None;
-  if is_recursive && not !rec_needed
-  && Warnings.is_active Warnings.Unused_rec_flag then begin
+  if is_recursive && not !rec_needed then begin
```

Comment:
It's rather non-intuitive that removing `Warnings.is_active ...` would make the warning more scope-aware. Why is this the correct change? Is the behavior correct because `prerr_warning` is a no-op when the warning is not in fact active? Does this mean that another fix would be to move `Builtin_attributes.warning_scope ... pvb_attribtes` just above the `if` test?

Reply:
> Is the behavior correct because `prerr_warning` is a no-op when the warning is not in fact active?

Yes.

> Does this mean that another fix would be to move `Builtin_attributes.warning_scope ...` pvb_attribtes just above the if test?

Yes.
I'd leave the  `if is_recursive` out (like the current PR does) though, only adding the `if Warnings.is_active ...` inside the warning scope. But really, that's the first thing that `prerr_warning` does, so...

Reply:
I think that the current code is okay -- no need to check twice -- but that the API is poor; I would expect `prerr_warning` to print *in any case*, and a different function `emit_if_enabled` or whatever to have a conditional behavior.

Reply:
There is something odd still, the rest of the codebase often checks `is_active` before `prerr_warning`

```ocaml
bytecomp/simplif.ml-569-      if ap.ap_should_be_tailcall
bytecomp/simplif.ml-570-      && not is_tail
bytecomp/simplif.ml:571:      && Warnings.is_active Warnings.Expect_tailcall
bytecomp/simplif.ml-572-        then Location.prerr_warning ap.ap_loc Warnings.Expect_tailcall;
[...]
typing/parmatch.ml:1944:      if Warnings.is_active Warnings.All_clauses_guarded then
typing/parmatch.ml-1945-        Location.prerr_warning loc Warnings.All_clauses_guarded
typing/parmatch.ml-1946-    end ;
[...]
typing/parmatch.ml:1965:            if Warnings.is_active (Warnings.Partial_match "") then begin
typing/parmatch.ml-1966-              let errmsg =
typing/parmatch.ml-1967-                try
typing/parmatch.ml-1968-                  let buf = Buffer.create 16 in
typing/parmatch.ml-1969-                  let fmt = Format.formatter_of_buffer buf in
typing/parmatch.ml-1970-                  Printpat.top_pretty fmt v;
typing/parmatch.ml-1971-                  if do_match (initial_only_guarded casel) [v] then
typing/parmatch.ml-1972-                    Buffer.add_string buf
typing/parmatch.ml-1973-                      "\n(However, some guarded clause may match this value.)";
typing/parmatch.ml-1974-                  if contains_extension v then
typing/parmatch.ml-1975-                    Buffer.add_string buf
typing/parmatch.ml-1976-                      "\nMatching over values of extensible variant types \
typing/parmatch.ml-1977-                         (the *extension* above)\n\
typing/parmatch.ml-1978-                    must include a wild card pattern in order to be exhaustive."
typing/parmatch.ml-1979-                  ;
typing/parmatch.ml-1980-                  Buffer.contents buf
typing/parmatch.ml-1981-                with _ ->
typing/parmatch.ml-1982-                  ""
typing/parmatch.ml-1983-              in
typing/parmatch.ml-1984-                Location.prerr_warning loc (Warnings.Partial_match errmsg)
typing/parmatch.ml-1985-            end;
[...]
typing/printtyp.ml:1683:  if Warnings.(is_active @@ Erroneous_printed_signature "")
typing/printtyp.ml-1684-  && Conflicts.exists ()
typing/printtyp.ml-1685-  then begin
typing/printtyp.ml-1686-    let conflicts = Format.asprintf "%t" Conflicts.print in
typing/printtyp.ml-1687-    Location.prerr_warning (Location.in_file sourcefile)
typing/printtyp.ml-1688-      (Warnings.Erroneous_printed_signature conflicts);
typing/printtyp.ml-1689-    Warnings.check_fatal ()
typing/printtyp.ml-1690-  end;
[...]
typing/typecore.ml-957-      if closed = Closed
typing/typecore.ml:958:      && Warnings.is_active (Warnings.Non_closed_record_pattern "")
typing/typecore.ml-959-      then begin
typing/typecore.ml-960-        let undefined = ref [] in
typing/typecore.ml-961-        for i = 0 to Array.length all - 1 do
typing/typecore.ml-962-          if not defined.(i) then undefined := all.(i).lbl_name :: !undefined
typing/typecore.ml-963-        done;
typing/typecore.ml-964-        if !undefined <> [] then begin
typing/typecore.ml-965-          let u = String.concat ", " (List.rev !undefined) in
typing/typecore.ml-966-          Location.prerr_warning loc (Warnings.Non_closed_record_pattern u)
typing/typecore.ml-967-        end
typing/typecore.ml-968-      end
```

Sometimes the test is used to save work (or to insert a callback to be used in a different context), but often it is just guarding the `prerr_warning` itself. There definitely seems to be an API issue to me.

Reply:
Yep, I know. "What can you do?"

I agree that there are various places where it's indeed saving work (parmatch being one of them for example IIRC).

But the "API" problem has been here forever, it hasn't appeared with Armael's changes (if that's what you had in mind).